Make RateLimiter::GetTotalPendingRequest() non pure virtual for backward compability

hx235 opened the pull request:
Context/Summary:
https://github.com/facebook/rocksdb/pull/8890 added a public API `RateLimiter::GetTotalPendingRequest()` but mistakenly marked it as pure virtual, forcing RateLimiter's derived classes to implement this function and breaking backward compatibility.

This PR makes `RateLimiter::GetTotalPendingRequest()` as non-pure virtual method by providing a trivial implementation in rate_limiter.h

Test:
Passing existing tests

pdillinger reviewed (changes requested):
This line in the HISTORY doesn't fully make sense anymore:

HISTORY.md:* Add a public API RateLimiter::GetTotalPendingRequests() for the total number of requests that are pending for bytes in the rate limiter.

It appears this function is not used by RocksDB internally but is provided by RocksDB built-in rate limiter implementations for use by user code.

And relevant to that: I suggest making the function return a more distinctive "not implemented" value such as -1 and also "assert(false)" for misuse detection in debug builds.

pdillinger reviewed (commented):
Actually, I think the most "accurate" way to add this feature is to not have this function in RateLimiter but put it in BuiltinRateLimiter derived from RateLimiter, and have NewGenericRateLimiter return BuiltinRateLimiter. But I'm not sure it's worth the extra complexity to be "accurate" with our types. (Other opinions welcome.)

hx235 commented:
Update:
- Updated HISTORY.md/trivial implementation of `RateLimiter::GetTotalPendingRequests()`  as suggested

ajkr commented:
I am guessing it is reasonable for a non-builtin `RateLimiter` (such as user wrappers, which we may see soon) to implement this function.

I thought a more traditional way to implement this would be to return a `Status::NotSupported()` by default, and overriders can assign the result to a pointer argument and return `Status::OK()`. Then callers can write code that uses the result when available or does something else when it's not supported in their `RateLimiter` object.

pdillinger commented:
It could be called ExtendedRateLimiter or RateLimiterV2 or something.

I'm not a big fan of Status for things that aren't subject to external factors, but ok in a case like this.

pdillinger commented on include/rocksdb/rate_limiter.h:
```diff
@@ -91,7 +91,11 @@ class RateLimiter {
 
   // Total # of requests that are pending for bytes in rate limiter
```
We need more description in here like

    For convenience, this function is implemented by the RateLimiter returned by NewGenericRateLimiter but is not required by RocksDB. The default implementation indicates "not supported."

pdillinger reviewed (approved):
Otherwise LGTM

pdillinger reviewed (commented):
Ideally, we should also have a smoke test with a custom RateLimiter (in test code) not implementing GetTotalPendingRequests, to hopefully detect if we start calling the function.

hx235 commented:
Update:
- Added comment for `RateLimiter::GetTotalPendingRequests()` as suggested
- Added a smoke test to verify the optional `RateLimiter::GetTotalPendingRequests()` is not required for basic RocksDB operation

hx235 commented:
Note: haven't imported the latest PR to fbsource since there is an error "An error occurred while syncing the pull request data from GitHub; please try again in a few minutes."

pdillinger commented:
LGTM :)

hx235 commented:
Update
- Removed unnecessary semicolon after member function definition
- Added key word "override" for overriding virtual destructor

ajkr commented on db/db_test.cc:
```diff
@@ -3937,6 +3937,56 @@ TEST_F(DBTest, DISABLED_RateLimitingTest) {
+  int64_t GetBytesPerSecond() const override { return 0; }
+};
+
+// To test that customed rate limiter not implementing optional APIs (e.g,
```
Honestly I find this solution strange (now there's an extension API that we're not supposed to call? Seems unnecessarily restrictive.). I hope in the future we can just follow the convention of returning Status::NotSupported.

hx235 replied:
Discussed offline
